### PR TITLE
Support for MCTS Beneficiary Updates

### LINKED ISSUE #113: NMS.GEN.MCTS.005 - Ability to modify phone number of kilkari beneficiary via UI or CSV
The NMS system shall provide means to modify the MSISDN of a Kilkari Beneficiary by providing new MSISDN information along with existing MCTS id using CSV upload or MDS UI. However, this procedure is not allowed if the number is already being used by another Active Kilkari beneficiary for same pack.
A history/change table shall be created with old and new number for the beneficiary whose MSISDN number is being changed.

The language mapping for the beneficiary does not change due to change in MSISDN number.

### LINKED ISSUE #114: NMS.GEN.MCTS.006 - Ability to modify kilkari beneficiary address via UI or CSV
The NMS system shall provide means to modify the address/location of a Kilkari Beneficiary by providing new address/location information along with existing MCTS ID using CSV upload or MDS UI.

A history/change table shall be created with old and new address/location for the existing beneficiary.
The language mapping for the beneficiary does not change due to change in location information.

### LINKED ISSUE #115: NMS.GEN.MCTS.007 - Ability to modify kilkari beneficiary LMP/DOB via UI or CSV
The NMS system shall provide means to modify the LMP/DOB of a Kilkari Beneficiary by providing new LMP/DOB information along with existing MCTS ID using CSV upload or MDS UI. The updated information shall be used to provide new schedule for the beneficiary.

A history/change table shall be created with old and new LMP/DOB for the existing beneficiary.

### LINKED ISSUE #116: NMS.GEN.MCTS.008 - Ability to modify kilkari beneficiary mother/child status via UI or CSV
The NMS system shall provide means to update mother/child status of a Kilkari Beneficiary using CSV upload or MDS UI. The following information status fields shall be available in internal database of NMS:
- Miscarriage/Abortion
- Still birth
- Child death
- Maternal Death

### LINKED ISSUE #162: NMS.KK.ACCESS.010 - Auto cancellation from MCTS upload
Any of the events listed in #116  NMS.GEN.MCTS.008 related to events for mother/child shall lead to stopping of Kilkari service (subscription state = Deactivated).

The service shall stop immediately if maternal death, infant death, miscarriage or abortion is reported via MCTS

### LINKED ISSUE #358: Reactivate/create new subscription if an updated reference date puts subscriber in a valid subscription window
Clarification from BBC:

If user has a completed pack, but the date change would imply that the pack would still be active, should we reactivate?
GF: No need to reactivate as the pack has already been completed.
o    Saurabh: In case of user gets her LMP/DOB preponed after completion of OBD cycle and new dates falls into OBD cycle again, we should activate her pack according to fresh LMP/DOB.

### PR DESCRIPTION
**Pull request by llavoie**
Fixes #113 - Ability to modify phone number of kilkari beneficiary via UI or CSV
Fixes #114 - Ability to modify kilkari beneficiary address via UI or CSV
Fixes #115 - Ability to modify kilkari beneficiary LMP/DOB via UI or CSV
Fixes #116 - Ability to modify kilkari beneficiary mother/child status via UI or CSV
Fixes #162 - Auto cancellation from MCTS upload
Fixes #358 - Reactivate/create new subscription if an updated reference date puts subscriber in a valid subscription window

For 113-116 I've omitted the change logging aspect because @kamknitt is working on that. I've also created tickets for a couple of small TODOs.

### CODE REVIEW

**Comment by larubbio** [comment]
@frankhuster Can you review this today?  I think once this is in all the issues labeled FT will be complete.

**Comment by frankhuster** [comment]
I'll get right on it now.

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/SubscriberService.java [hunk 1]
```diff
@@ -30,6 +41,18 @@
...
     /**
+     * Update MSISDN for subscriber. If the new MSISDN is already in use by a different MCTS beneficiary for the same
+     * subscription pack, the update will be rejected.
+     * @param subscriber Existing subscriber object
+     * @param beneficiary The MCTS beneficary (mother or child) to update
+     * @param newMsisdn The new MSISDN for the subscriber
+     */
+    void updateMsisdnForSubscriber(Subscriber subscriber, MctsBeneficiary beneficiary, Long newMsisdn);
+
+    Subscription updateOrCreateMctsSubscriber(MctsBeneficiary beneficiary, Long msisdn, DateTime referenceDate,
```
Comment (frankhuster) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/SubscriberService.java]: No javadoc?

File: testing/src/test/java/org/motechproject/nms/testing/it/kilkari/MctsBeneficiaryImportServiceBundleIT.java [hunk 2]
```diff
@@ -286,22 +294,73 @@ public void testImportMotherAndChildSameMsisdn() throws Exception {
...
+
+        Subscription subscription = subscriptions.iterator().next();
+        assertEquals(SubscriptionStatus.ACTIVE, subscription.getStatus());
+
+        State expectedState = stateDataService.findByCode(21L);
+        District expectedDistrict = districtService.findByStateAndCode(expectedState, 3L);
+
+        assertMother(subscriber, "1234567890", getDateTime(lmpString), "Shanti Ekka", expectedState, expectedDistrict);
+
+        // import record for same mother with Abortion set to "Spontaneous" -- her subscription should be deactivated
```
Comment (llavoie) [inline testing/src/test/java/org/motechproject/nms/testing/it/kilkari/MctsBeneficiaryImportServiceBundleIT.java]: Copy-paste error in this comment -- should say Outcome_Nos set to 0

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryImportServiceImpl.java [hunk 3]
```diff
@@ -376,34 +293,25 @@ public MctsMother retrieve(String value) {
...
-        mapping.put(MSISDN, new GetLong());
-        mapping.put(LMP, new Optional(new GetInstanceByString<DateTime>() {
+        mapping.put(MSISDN, MctsBeneficiaryUtils.MSISDN_BY_STRING);
+        mapping.put(LMP, new Optional(MctsBeneficiaryUtils.DATE_BY_STRING));
+        mapping.put(MOTHER_DOB, new Optional(MctsBeneficiaryUtils.DATE_BY_STRING));
+        mapping.put(ABORTION, new Optional(new GetInstanceByString<Boolean>() {
+             @Override
+             public Boolean retrieve(String value) {
+                 String trimmedValue = value.trim();
+                 return "Spontaneous".equals(trimmedValue) || "MTP<12 Weeks".equals(trimmedValue) ||
```
Comment (frankhuster) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryImportServiceImpl.java]: case sensitive?

Comment (frankhuster) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryImportServiceImpl.java]: exact or fuzzy matches?

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUpdateServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+    @Override
+    public void updateBeneficiaryData(Reader reader) throws IOException {
+        BufferedReader bufferedReader = new BufferedReader(reader);
+
+        CsvMapImporter csvImporter = new CsvImporterBuilder()
+                .setProcessorMapping(getProcessorMapping())
+                .setPreferences(CsvPreference.STANDARD_PREFERENCE)
+                .createAndOpen(bufferedReader);
+        try {
```
Comment (frankhuster) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUpdateServiceImpl.java]: We're rejecting the entire file at the first error, right?

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUpdateServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+        if (containsLocationUpdate(record)) {
+            // validate and set location
+            try {
+                MctsBeneficiaryUtils.setLocationFields(locationService.getLocations(record), beneficiary);
+                beneficiaryDataService.update(beneficiary);
+            } catch (InvalidLocationException le) {
+                LOGGER.error(le.toString());
+
+               // TODO: How do we log this failed update? SubscriptionError doesn't seem like the right mechanism.
```
Comment (frankhuster) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUpdateServiceImpl.java]: Why not have a `LocationError` entity similar to `SubscriptionError`?

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUpdateServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+        // Third, update the beneficiary's subscription date/status based on the new reference date if provided
+
+        // if subscriber is null (and we have the necessary info), or we have a new reference date, create/update the
+        // subscription for this beneficiary
+        if ((subscriber == null) && (newMsisdn != null) && (newReferenceDate != null)) {
+            // create a new subscription if the beneficiary's LMP/DOB indicates that a subscription should be created
+
+            subscriberService.updateOrCreateMctsSubscriber(beneficiary, newMsisdn, newReferenceDate, packType);
+            return;
+        } else if (newReferenceDate != null) {
```
Comment (frankhuster) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUpdateServiceImpl.java]: simplify and remove the `else` statement for readability?

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUpdateServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+        }
+        if (beneficiary == null) {
+            beneficiary = mctsChildDataService.findByBeneficiaryId(stateMctsId);
+        }
+
+        return beneficiary;
+    }
+
+    private boolean containsLocationUpdate(Map<String, Object> record) {
+        return ((record.get(STATE) != null) || (record.get(DISTRICT) != null) || (record.get(TALUKA) != null) || //NO CHECKSTYLE BooleanExpressionComplexity
```
Comment (frankhuster) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUpdateServiceImpl.java]: [nitpick alert]
How about:
```
        if (record.get(STATE) != null) {
            return true;
        }
        if (record.get(DISTRICT) != null) {
            return true;
        }
        if (record.get(STATE) != null) {
            return true;
        }
        if (record.get(TALUKA) != null) {
            return true;
        }
        if (record.get(HEALTH_BLOCK) != null) {
            return true;
        }
        if (record.get(SUBCENTRE) != null) {
            return true;
        }
        if (record.get(CENSUS_VILLAGE) != null) {
            return true;
        }
        
        return false;
```
to get rid of the `//NO CHECKSTYLE`

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUtils.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            throw new InvalidLocationException("Missing mandatory district field");
+        }
+
+        beneficiary.setState((State) locations.get(STATE));
+        beneficiary.setDistrict((District) locations.get(DISTRICT));
+        beneficiary.setTaluka((Taluka) locations.get(TALUKA));
+        beneficiary.setHealthBlock((HealthBlock) locations.get(HEALTH_BLOCK));
+        beneficiary.setHealthFacility((HealthFacility) locations.get(PHC));
+        beneficiary.setHealthSubFacility((HealthSubFacility) locations.get(SUBCENTRE));
+        beneficiary.setVillage((Village) locations.get(CENSUS_VILLAGE + NON_CENSUS_VILLAGE));
```
Comment (frankhuster) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUtils.java]: Are you constructing a key by concatenation?

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/SubscriberServiceImpl.java [hunk 9]
```diff
@@ -35,6 +64,35 @@ public Subscriber getSubscriber(long callingNumber) {
...
+                query.setClass(Subscriber.class);
+                Long id = beneficiary.getId();
+                ForwardQueryResult fqr = (ForwardQueryResult) query.execute(id, id);
+                if (fqr.isEmpty()) {
+                    return null;
+                }
+                if (fqr.size() == 1) {
+                    return (Subscriber) fqr.get(0);
+                }
+                throw new IllegalStateException("More than one row returned!");
```
Comment (frankhuster) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/SubscriberServiceImpl.java]: Add the ID in the exception message to help with debugging?

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/SubscriberServiceImpl.java [hunk 10]
```diff
@@ -67,6 +125,113 @@ public void update(Subscriber subscriber) {
...
+    @Override
+    public Subscription updateOrCreateMctsSubscriber(MctsBeneficiary beneficiary, Long msisdn, DateTime referenceDate,
+                                                     SubscriptionPackType packType) {
+        District district = beneficiary.getDistrict();
+        Language language = (Language) districtDataService.getDetachedField(district, "language");
+        Subscriber subscriber = getSubscriber(msisdn);
+
+        SubscriptionPack pack = subscriptionPackDataService.byType(packType);
+
+        // TODO: Handle the case in which the MCTS beneficiary already exists but with a different phone number
```
Comment (frankhuster) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/SubscriberServiceImpl.java]: Should we have a ticket number for that?
Are we waiting for input before we can proceed?

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/SubscriberServiceImpl.java [hunk 11]
```diff
@@ -67,6 +125,113 @@ public void update(Subscriber subscriber) {
...
+
+            MctsBeneficiary existingBeneficiary = (packType == SubscriptionPackType.PREGNANCY) ? subscriber.getMother() :
+                    subscriber.getChild();
+
+            if (existingBeneficiary == null) {
+                // there's already an IVR-originated subscription for this MSISDN
+                subscriptionErrorDataService.create(
+                        new SubscriptionError(msisdn, SubscriptionRejectionReason.ALREADY_SUBSCRIBED, packType));
+
+                // TODO: Do we just reject the subscription request, or do we update the subscriber record with MCTS data?
```
Comment (frankhuster) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/SubscriberServiceImpl.java]: need ticket#?

File: testing/src/test/java/org/motechproject/nms/testing/it/kilkari/MctsBeneficiaryUpdateServiceBundleIT.java [hunk 12]
```diff
@@ -0,0 +1,525 @@
...
+
+    private Reader read(String resource) {
+        return new InputStreamReader(getClass().getClassLoader().getResourceAsStream(resource));
+    }
+
+    private String getDateString(DateTime date) {
+        return date.toString("dd-MM-yyyy");
+    }
+
+}
```
Comment (frankhuster) [inline testing/src/test/java/org/motechproject/nms/testing/it/kilkari/MctsBeneficiaryUpdateServiceBundleIT.java]: How about testing some failure cases and making sure you're getting the right errors/exceptions/alerts?

**Comment by frankhuster** [comment]
Few nitpicks
Confused by the 
`beneficiary.setVillage((Village) locations.get(CENSUS_VILLAGE + NON_CENSUS_VILLAGE));`
line.

Other than that, LGTM!

**Comment by larubbio** [comment]
@llavoie Is it ok if this is merged and then the nitpicks get picked up in a later commit?  I want to get this in so the remaining tests are unblocked.

**Comment by larubbio** [comment]
ooops.  I was thinking about merging and hovering over the button, and then .......

**Comment by llavoie** [comment]
That's fine. :) I'll address the feedback in a separate checkin tomorrow.

-----Original Message-----

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryImportServiceImpl.java (hunk 3 above)
Comment (llavoie) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryImportServiceImpl.java]: I have no idea. For now I'm assuming exact match (with the possible exception of leading/trailing spaces). But we haven't seen any real data yet.

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUpdateServiceImpl.java (hunk 4 above)
Comment (llavoie) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUpdateServiceImpl.java]: In general, we shouldn't. There are a few places (here as well as in the FLW import code) that this should be cleaned up.

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUtils.java (hunk 8 above)
Comment (llavoie) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/MctsBeneficiaryUtils.java]: This is @koshalt 's code, I'm not sure I really understand how it works.

File: testing/src/test/java/org/motechproject/nms/testing/it/kilkari/MctsBeneficiaryUpdateServiceBundleIT.java (hunk 12 above)
Comment (llavoie) [inline testing/src/test/java/org/motechproject/nms/testing/it/kilkari/MctsBeneficiaryUpdateServiceBundleIT.java]: Aricent is doing that. :)

**Pull request merged**